### PR TITLE
Capture UX telemetry

### PR DESCRIPTION
**Pull request by harrdou**
Capture custom telemetry for user interactions

### CODE REVIEW

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py [hunk 1]
```diff
@@ -97,12 +98,13 @@ def AuthenticateOutOfBand(self, requestParameters):
         authenticationService = CdiUtil.bean(AuthenticationService)
         authenticationProtectionService = CdiUtil.bean(AuthenticationProtectionService)
 
-        if requestParameters.containsKey("oob:cancel"):
-            return False
-
+        session = identity.getSessionId()
```
Comment (wmoussa-gc) [inline gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:101]: Im sure it is a non-issue, but this is really a session object not just an id, correct?

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py [hunk 2]
```diff
@@ -114,6 +116,10 @@ def AuthenticateOutOfBand(self, requestParameters):
             addMessage("oob:resend", FacesMessage.SEVERITY_INFO, "sic.newCode")
 
         if codeExpired or requestParameters.containsKey("oob:resend"):
+            telemetry["result"] = "failed"
+            telemetry["reason"] = "code expired" if codeExpired else "new code requested"
+            self.telemetryClient.trackEvent("OOB Authentication", telemetry, {"duration": duration})
```
Comment (wmoussa-gc) [inline gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:121]: would be good to mention what time unit this duration is, seconds?

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py [hunk 3]
```diff
@@ -130,14 +136,19 @@ def AuthenticateOutOfBand(self, requestParameters):
         if enteredCode == identity.getWorkingParameter("oobCode"):
             facesMessages.clear()
             updateNeeded = False
+            telemetry["result"] = "sucess"
```
Comment (wmoussa-gc) [inline gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:139]: typo

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py [hunk 4]
```diff
@@ -130,14 +136,19 @@ def AuthenticateOutOfBand(self, requestParameters):
...
 
             if contact is not None: # Registration
+                telemetry["step"] = "code verification"
```
Comment (wmoussa-gc) [inline gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:142]: isn't this a step for both oob registration and auth?

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py [hunk 5]
```diff
@@ -154,14 +165,22 @@ def AuthenticateOutOfBand(self, requestParameters):
 
             return authenticationService.authenticate(userId)
         else:
+            telemetry["result"] = "failed"
+            telemetry["reason"] = "invalid code"
+            if contact is not None:
+                telemetry["step"] = "code verification"
```
Comment (wmoussa-gc) [inline gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:171]: isn't a step for both, registration and auth?

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py [hunk 6]
```diff
@@ -686,6 +732,8 @@ def getNextStep(self, configurationAttributes, requestParameters, step):
         session = identity.getSessionId()
         rpConfig = self.rputils.getRPConfig(session)
         provider = identity.getWorkingParameter("provider")
+        if provider == "gckeyregister": # Hack
```
Comment (wmoussa-gc) [inline gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py:735]: was this hack needed for this PR or could it be pushed into a new PR and explain the reason?

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py [hunk 7]
```diff
@@ -284,11 +287,13 @@ def prepareForStep(self, configurationAttributes, requestParameters, step):
                     break
 
         if step in {self.STEP_1FA, self.STEP_COLLECT, self.STEP_TOTP_REGISTER, self.STEP_TOTP}: # Passport
-            
             passportOptions = {"ui_locales": uiLocales, "exp" : int(time.time()) + 60}
 
             if step in {self.STEP_1FA, self.STEP_COLLECT}:
                 provider = identity.getWorkingParameter("provider")
+                telemetry = {"sessionId" : session.getId(), "provider": provider}
+                if provider == "gckeyregister": # Hack
```
Comment (wmoussa-gc) [inline gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py:295]: is it needed for this PR purpose?

**Review by wmoussa-gc — COMMENTED** [review]
Thanks Doug for pushing this through. Could we prepare a sidecar sheet that list all the possible telemetry events, classification and metrics that you have coded? I don't expect that sheet to be maintained but since we're launching this new logging would be good to have a review outside the code

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 1 above)
Comment (harrdou) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:101]: Funny how that's the first thing everybody notices. I know of two  standout examples of poor naming choices in the Gluu source code. [That's one](https://github.com/GluuFederation/oxAuth/blob/master/Server/src/main/java/org/gluu/oxauth/model/common/SessionId.java).

The other is this pair of accessors [here](https://github.com/GluuFederation/fido2/blob/2b97bc5b1799bc3f4a5eb23d6e1731188406365d/server/src/main/java/org/gluu/fido2/model/auth/AuthData.java#L96-L102) which actually set or return the signature algorithm, not the key type.

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py (hunk 7 above)
Comment (harrdou) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py:295]: @stephenyates-gc  asked for it.

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 4 above)
Comment (harrdou) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:142]: Yes. Autentication is tracked [here](https://github.com/sign-in-canada/Acceptance-Platform/blob/ebdb3eeb376670782dd4b84b5143b459028957ae/gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py#L151).

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 5 above)
Comment (harrdou) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:171]: Yes. https://github.com/sign-in-canada/Acceptance-Platform/blob/ebdb3eeb376670782dd4b84b5143b459028957ae/gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py#L170-L174

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 3 above)
Comment (harrdou) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py]: Good catch. Thanks.

https://github.com/sign-in-canada/Acceptance-Platform/commit/993a22dc7ba0631f88d4021b57789a0d13e6012a

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 2 above)
Comment (harrdou) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:121]: Are you suggesting a name change (e.g. durationInSeconds) or just some documentation? This is essentially the think time, i.e. the time the user needed to complete that particular task.

I'm not enamored with the variable name "telemetry" which I've used throughout either. It is really the collection of  event properties. Maybe should be "eventProperties" instead?

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 4 above)
Comment (wmoussa-gc) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:142]: I meant shouldn't be outside the if statement (i.e injected in line 140)

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 2 above)
Comment (wmoussa-gc) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:121]: the name change will help when looking at the event log and the metric unit. as for telemetry, yeah it is quite generic but I think good enough to understand it is a temporary object holder for observabiity

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 5 above)
Comment (wmoussa-gc) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:171]: in that case I think line 171 should site outside the if/else block

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py (hunk 6 above)
Comment (wmoussa-gc) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/person_authentication/SignInCanada.py:735]: oh I see now, thanks for sharing

**Review by wmoussa-gc — APPROVED** [review]
lgtm

File: gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py (hunk 5 above)
Comment (harrdou) [reply gluu-server/opt/gluu/jetty/oxauth/custom/scripts/sic/oob.py:171]: It's the 2nd of 2 steps for registration, but the only step for authentication. No harm in including it for both however.

**Comment by shippy6969** [comment]
I did not test the telemetry but did successfully test the key functionality of the MFA plus perform regression testing.

**Comment by shippy6969** [comment]
Where do I approve Doug please?

**Pull request merged**